text-emphasis: Do not paint emphasis marks on unicode punctuations

Ahmad-S792 opened the pull request:
#### ea1eac388c7648b2a257c99e10a3cf0504e1ec99
<pre>
text-emphasis: Do not paint emphasis marks on unicode punctuations

<a href="https://bugs.webkit.org/show_bug.cgi?id=285405">https://bugs.webkit.org/show_bug.cgi?id=285405</a>
<a href="https://rdar.apple.com/142387538">rdar://142387538</a>

Reviewed by Darin Adler.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium and Web Specification [1]:

[1] <a href="https://drafts.csswg.org/css-text-decor-3/#text-emphasis-style-property">https://drafts.csswg.org/css-text-decor-3/#text-emphasis-style-property</a>

&quot;Punctuation--specifically, any characters that belong to the Unicode P* general category...&quot;

and was clarified in: <a href="https://github.com/w3c/csswg-drafts/issues/839">https://github.com/w3c/csswg-drafts/issues/839</a>

Inspired by: <a href="https://source.chromium.org/chromium/chromium/src/+/dba2b53e93f71fa2df02d72a42a42371a63c695b">https://source.chromium.org/chromium/chromium/src/+/dba2b53e93f71fa2df02d72a42a42371a63c695b</a> and
<a href="https://phabricator.services.mozilla.com/D219099">https://phabricator.services.mozilla.com/D219099</a>

This patch adds unicode punctuations list, which should not receive `text emphasis`
painting / styling as well those which should (matching Firefox implementation).

Despite aligning WebKit with Firefox implementation, we are not matching expected output
to enable `text-emphasis-punctuation-3.html` but we are not regressing behavior as well wit
this change.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::canReceiveTextEmphasis):
* Source/WTF/wtf/Unicode/CharacterNames.h: Sorted alphabetically as well.

Canonical link: <a href="https://commits.webkit.org/288476@main">https://commits.webkit.org/288476@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4e92a2575b262c6ab543720769998cf136a7a81

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3155 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37832 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88610 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34547 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85624 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3243 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11114 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64991 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22733 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86585 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/2382 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75904 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45277 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2286 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30118 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33595 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/76502 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73369 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30848 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89987 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/82556 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10804 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7794 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/73422 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11027 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71724 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72647 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17969 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16882 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15594 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/2128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10756 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16228 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/104973 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10602 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25375 "Found 3 new JSC stress test failures: stress/json-stringify-inspector-check.js.no-llint, stress/substring-global-atom-cache.js.lockdown, wasm.yaml/wasm/stress/cc-int-to-int-cross-module-with-exception.js.default-wasm (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14078 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12376 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1254,10 +1254,15 @@ bool FontCascade::canReceiveTextEmphasis(char32_t c)
         return false;
 
+    // Punctuations
+    if (U_GET_GC_MASK(c) & (U_GC_PD_MASK | U_GC_PS_MASK | U_GC_PE_MASK | U_GC_PC_MASK | U_GC_PO_MASK | U_GC_PI_MASK | U_GC_PF_MASK))
```
```suggestion
    if (U_GET_GC_MASK(c) & U_GC_P_MASK)
```

Ahmad-S792 replied:
@darinadler - Done. :-)

darinadler commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1254,10 +1254,15 @@ bool FontCascade::canReceiveTextEmphasis(char32_t c)
 
+    // Punctuations
+    if (U_GET_GC_MASK(c) & U_GC_P_MASK)
+        return false;
```
I realized this can be combined with the checks above and should not be a separate if statement. Just or in `U_GC_P_MASK` with the other masks in the expression at the top of the function, which is already checking multiple categories. Adding punctuation doesn't need to be separate lines of code with a separate comment. Adding a comment would be OK, but one just about punctuation doesn't make a ton of sense.
```
... U_GC_CF_MASK | U_GC_P_MASK))
```

Ahmad-S792 replied:
Done. :-)

nt1m commented on LayoutTests/TestExpectations:
```diff
@@ -4211,8 +4211,6 @@ imported/w3c/web-platform-tests/css/css-text-decor/text-emphasis-position-under-
 imported/w3c/web-platform-tests/css/css-text-decor/text-emphasis-style-002.html [ ImageOnlyFailure ]
-imported/w3c/web-platform-tests/css/css-text-decor/text-emphasis-punctuation-1.html [ ImageOnlyFailure ]
-imported/w3c/web-platform-tests/css/css-text-decor/text-emphasis-punctuation-2.html [ ImageOnlyFailure ]
 imported/w3c/web-platform-tests/css/css-text-decor/text-emphasis-punctuation-3.html [ ImageOnlyFailure ]
```
There's a few allowed punctuation symbols fwiw: https://drafts.csswg.org/css-text-decor/#text-emphasis-style-property.

The 3rd test is testing precisely those exceptions. Can we address them?

<img width="828" alt="image" src="https://github.com/user-attachments/assets/1805d7f7-a033-47cb-b038-45e20727dd59" />

Ahmad-S792 replied:
Yes - I know. yes, can try it in separate PR?

nt1m replied:
Chrome doesn't implement these exceptions and the difference on that test is pretty obvious, many emphasis marks are not painted when they should be: https://wpt.fyi/analyzer?screenshot=sha1%3A647615a53bdd0ba063b31f87356b7866a3a4d92d&screenshot=sha1%3A8a8008e193af7b79b23ee4b9008b8c9354f81306

Firefox implements them properly though: https://wpt.fyi/analyzer?screenshot=sha1%3A85451987d4f4a0949dafc045df7f9402d9487f35&screenshot=sha1%3A4fc60ec3f66ef776258cb69220f91dc184151577

nt1m replied:
> Yes - I know. yes, can try it in separate PR?

I'd prefer doing it in the same PR. It's a regression otherwise if we stop painting emphasis marks where there used to be some. Here's Mozilla's commit in case it helps: https://phabricator.services.mozilla.com/D219099

Ahmad-S792 replied:
Raised - https://bugs.webkit.org/show_bug.cgi?id=285409

nt1m replied:
Can we try to fix it here? It really shouldn't be hard

darinadler replied:
I agree with @nt1m: We shouldn't unnecessarily break them and then fix them.

darinadler replied:
They worked until we added the punctuation exception. These are not two separate issues.

Ahmad-S792 replied:
Tackling in one PR, so I think comment is now resolved.

nt1m commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1250,10 +1250,11 @@ bool FontCascade::rightExpansionOpportunity(StringView stringView, TextDirection
-    if (U_GET_GC_MASK(c) & (U_GC_Z_MASK | U_GC_CN_MASK | U_GC_CC_MASK | U_GC_CF_MASK))
+    if (U_GET_GC_MASK(c) & (U_GC_Z_MASK | U_GC_CN_MASK | U_GC_CC_MASK | U_GC_CF_MASK | U_GC_P_MASK))
         return false;
 
```
```
if (U_GET_GC_MASK(c) & U_GC_P_MASK)
    return c == '#' || c == '%' || c == '&' || [...]
```

Implementing the exceptions should be something like this, but maybe @darinadler has a better idea.

I'd reuse the list from this Firefox change: https://phabricator.services.mozilla.com/D219099

darinadler replied:
If we have a performance problem we can make an ASCII table or a sorted array map. Generally a special case for ASCII works well, with one branch to range check and then a simple index into a 128-element array.

darinadler reviewed (changes requested):
We should not regress handling of the punctuation exceptions.

nt1m reviewed (commented):
I don't really know unicode-related code well enough to review this, so I'll leave it to someone else to approve.

Can you add a link to the spec before `canReceiveTextEmphasis` ? https://www.w3.org/TR/css-text-decor-3/#text-emphasis-style-property

darinadler commented on Source/WTF/wtf/unicode/CharacterNames.h:
```diff
@@ -153,6 +153,29 @@ constexpr UChar inhibitArabicFormShaping = 0x206C;
+constexpr UChar numberSign = 0x0023;
+constexpr UChar percentageSign = 0x0025;
+constexpr UChar ampersand = 0x0026;
+constexpr UChar commercialAt = 0x0040;
```
I don't think we benefit from having named constants for these ASCII characters.

Ahmad-S792 replied:
Removed - thanks! I was just trying to be consistent.

darinadler commented on Source/WTF/wtf/unicode/CharacterNames.h:
```diff
@@ -153,6 +153,29 @@ constexpr UChar inhibitArabicFormShaping = 0x206C;
+constexpr UChar fullwidthNumberSign = 0xFF03;
+constexpr UChar fullwidthPercentSign = 0xFF05;
+constexpr UChar fullwidthAmpersand = 0xFF06;
+constexpr UChar fullwidthCommericalAt = 0xFF20;
```
We started this file alphabetical. Not sure at what point we stopped doing that for new constants. Please sort alphabetically in with the constants we already have, and please also sort in the more recent additions that aren't alphabetical, which seem to start with `lineSeparator`.

It would also be OK to sort numerically like you did here, but they are part of a larger list and we want a single sort, not multiple.

Ahmad-S792 replied:
Sorted alphabetically now and also removed suggested ones.

darinadler commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1254,10 +1254,20 @@ bool FontCascade::canReceiveTextEmphasis(char32_t c)
+            || c == pilcrowSign || c == arabicIndicPerMileSign || c == arabicIndicPerTenThousandSign || c == aracbicPercentSign
+            || c == perMileSign || c == perMileSign || c == perTenThousandSign || c == tironianSignSet || c == reversedPilcrowSign
+            || c == swungDash || c == partAlternationMark ||  c == smallNumberSign || c == smallAmpersand || c == smallPercentSign
+            || c == smallCommericalAt || c == fullwidthNumberSign || c == fullwidthPercentSign || c == fullwidthAmpersand || c == fullwidthCommericalAt;
```
This is generally OK. We should research if ICU happens to have some kind of category for these that could prevent us from needing to write out a table, but I'm guessing maybe not. There are probably more efficient ways to implement this, but for now we can just assume this is efficient enough. There is at least one stray space, `sectionSign` is missing, `perMilleSign` is listed twice, _arabic_ is misspelled as _aracbic_, _commercial_ is misspelled as _commerical_, _mille_ is misspelled as _mile_, and _sign et_ is misspelled as _sign set_. It's too hard to see if this is correct if it's formatted like this so I suggest vertical, sorted formatting:
```suggestion
        return c == '#' || c == '%' || c == '&' || c == '@'
            || c == arabicIndicPerMilleSign
            || c == arabicIndicPerTenThousandSign
            || c == arabicPercentSign
            || c == fullwidthAmpersand
            || c == fullwidthCommercialAt
            || c == fullwidthNumberSign
            || c == fullwidthPercentSign
            || c == partAlternationMark
            || c == perMilleSign
            || c == perTenThousandSign
            || c == pilcrowSign
            || c == reversedPilcrowSign
            || c == sectionSign
            || c == smallAmpersand
            || c == smallCommercialAt
            || c == smallNumberSign
            || c == smallPercentSign
            || c == swungDash
            || c == tironianSignEt
       ;
```
Since _section sign_ was missing, and we didn't notice, we should look at the results of the test carefully since it's still failing and make sure we understand the failures are not omissions from this list.

Ahmad-S792 replied:
I reduced the failure in JSFiddle and it only happens when I add `number` sign or unicode next to `C` and then C gets two emphasis mark. https://jsfiddle.net/pshk6dma/1/

Additionally, we have `line spacing` (line height issue), which might be something coming from subpixel issue probably. Might need to check with Alan / Simon / Antti.

darinadler commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1254,10 +1254,20 @@ bool FontCascade::canReceiveTextEmphasis(char32_t c)
         || c == ugariticWordDivider || c == tibetanMarkIntersyllabicTsheg || c == tibetanMarkDelimiterTshegBstar)
         return false;
 
+    // Punctuations
```
"Punctuations" is poor grammar. Should be "punctuation" or "punctuation characters". But also that comment doesn't explain the exceptions list and doesn't seem super helpful. I think we either should omit the comment or write a better one. The list of exceptions comes from the specification so we might want a link.

Ahmad-S792 replied:
Removed comment post @nt1m suggestion to add web spec link top of function.

darinadler commented on Source/WTF/wtf/unicode/CharacterNames.h:
```diff
@@ -141,30 +169,26 @@ constexpr UChar zeroWidthJoiner = 0x200D;
-constexpr UChar nominalDigitShapes = 0x206F;
+constexpr char32_t aegeanWordSeparatorDot = 0x10101;
+constexpr char32_t aegeanWordSeparatorLine = 0x10100;
+constexpr char32_t ugariticWordDivider = 0x1039F;
```
I would prefer to leave these sorted in based on the character names, not the type names. Can do that with the sort command line tool by passing an argument in to sort by the 3rd field.

Ahmad-S792 replied:
Done. :-)

darinadler commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1248,16 +1248,41 @@ bool FontCascade::rightExpansionOpportunity(StringView stringView, TextDirection
         || c == ugariticWordDivider || c == tibetanMarkIntersyllabicTsheg || c == tibetanMarkDelimiterTshegBstar)
         return false;
 
+    if (U_GET_GC_MASK(c) & U_GC_P_MASK) {
```
I think we should check if U_GET_GC_MASK is a costly operation. I suspect it is and if so we can compute it only once:
```
auto mask = U_GET_GC_MASK(c);
...
if (mask & U_GC_P_MASK)
```

Ahmad-S792 replied:
Done - I extended it to first function as well.

Ahmad-S792 committed:
text-emphasis: Do not paint emphasis marks on unicode punctuations

https://bugs.webkit.org/show_bug.cgi?id=285405
rdar://142387538

Reviewed by Darin Adler.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium and Web Specification [1]:

[1] https://drafts.csswg.org/css-text-decor-3/#text-emphasis-style-property

"Punctuation--specifically, any characters that belong to the Unicode P* general category..."

and was clarified in: https://github.com/w3c/csswg-drafts/issues/839

Inspired by: https://source.chromium.org/chromium/chromium/src/+/dba2b53e93f71fa2df02d72a42a42371a63c695b and
https://phabricator.services.mozilla.com/D219099

This patch adds unicode punctuations list, which should not receive `text emphasis`
painting / styling as well those which should (matching Firefox implementation).

Despite aligning WebKit with Firefox implementation, we are not matching expected output
to enable `text-emphasis-punctuation-3.html` but we are not regressing behavior as well wit
this change.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::canReceiveTextEmphasis):
* Source/WTF/wtf/Unicode/CharacterNames.h: Sorted alphabetically as well.

Canonical link: https://commits.webkit.org/288476@main
